internal/pipe/milestone/milestone.go

```diff
@@ -0,0 +1,132 @@
+package milestone
+
+import (
+	"github.com/apex/log"
+	"github.com/goreleaser/goreleaser/internal/client"
+	"github.com/goreleaser/goreleaser/internal/git"
+	"github.com/goreleaser/goreleaser/internal/pipe"
+	"github.com/goreleaser/goreleaser/internal/tmpl"
+	"github.com/goreleaser/goreleaser/pkg/config"
+	"github.com/goreleaser/goreleaser/pkg/context"
+)
+
+const defaultNameTemplate = "{{ .Tag }}"
+
+// Pipe for milestone.
+type Pipe struct{}
+
+func (Pipe) String() string {
+	return "milestones"
+}
+
+// Default sets the pipe defaults.
+func (Pipe) Default(ctx *context.Context) error {
+	if len(ctx.Config.Milestones) == 0 {
+		ctx.Config.Milestones = append(ctx.Config.Milestones, config.Milestone{})
+	}
+
+	for i := range ctx.Config.Milestones {
+		milestone := &ctx.Config.Milestones[i]
+
+		if milestone.NameTemplate == "" {
+			milestone.NameTemplate = defaultNameTemplate
+		}
+
+		// nolint: exhaustive
+		switch ctx.TokenType {
+		case context.TokenTypeGitLab:
+			{
+				if milestone.GitLab.Name == "" {
+					repo, err := git.ExtractRepoFromConfig()
+					if err != nil {
+						return err
+					}
```

Comment:
repeated a couple of times, maybe move to outside the switch?

Reply:
Sure thing -- I'm guessing `internal/pipe/release/release.go` will need the same treatment, but I'll leave it alone in this changeset.

Reply:
Hmm it might not be so clean since it will still need the conditionals for error handling (e.g. only error when the context token matches the VCS type and its not configured). Willing to adjust as necessary though if you have ideas!

Reply:
Would it help to maybe leave the GitLab/GitHub/Gitea logic entirely outside of the `pipe/milestone` and deal with the differences in `client.CloseMilestone`?

Also if the config is unified, as I suggested in my other comment, so that may help too?

Reply:
Updated to single `repo` configuration in e95effc3db84a45d4a067a428490945baf7057df 👍